MANTA-4874 tooling and docs to migrate mantav1 -> mantav2

trentm opened the pull request:
https://github.com/joyent/manta/blob/MANTA-4874/docs/operator-guide/mantav2-migration.md

During code review, use the following for the first step of migration to get the experimental images:

```
# MANTA-4811 and MANTA-4874
sdcadm_image=$(updates-imgadm -C experimental list name=sdcadm -j | json -c '~this.tags.buildstamp.indexOf("PR-60-")' -a uuid | tail -1)
manta_deployment_image=$(updates-imgadm list -C experimental name=mantav2-deployment version=~PR-60- --latest -H -o uuid)

sdcadm self-update -C experimental $sdcadm_image
sdcadm post-setup manta -C experimental -i $manta_deployment_image
```

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,595 @@
+- Are there old GC instruction files we should look into? Should `mako_gc.sh`
+  on the current makos handle those? How should we check?
+- Validate the "Accelerated GC" flush & disable steps above.
+
```
I believe the steps to follow are:
1. disable all garbage-collector SMF services to allow inflight instructions to drain
```
manta-oneach -s garbage-collector 'svcadm disable garbage-collector'
```
Wait 5 minutes and check that all instructions have drained:
```
manta-oneach -s garbage-collector 'du --inodes /var/spool/manta_gc/mako/* | sort -n | tail -3'
```
The file counts should all be 1 (the subdirectory itself).

2. **[For Manta deployment using "feeder" service only]** After 5 minutes, check that the feeder zone also has no inflight instructions:
```
du --inodes /var/spool/manta_gc/mako/* | sort -n | tail -3
```
The file counts should all be 1 (the subdirectory itself).

3. Before upgrading a storage zone to the v2 image, check that its instruction directory is empty:
**For Manta deployment using feeder service**
```
du --inodes /var/spool/manta_gc/instructions
```
The file count should be exactly 1 (the directory itself).

**For Manta deployment without feeder service**
```
minfo /poseidon/stor/manta_gc/mako/$(json -f /var/tmp/metadata.json -ga MANTA_STORAGE_ID) | grep result
```
The result set size should be exactly 1 (the directory itself).

trentm replied:
Thanks, I've added these instructions to https://github.com/joyent/manta/blob/MANTA-4874/docs/operator-guide/mantav2-migration.md#accelerated-gc

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,595 @@
+Select the driver DC. Results from subsequent phases need to be collected
+in one place. Therefore, if this Manta region has multiple DCs, then you
+will be asked to choose one of them on which to coordinate. This is called
+the "driver DC". Any of the DCs in the Manta region will suffice.
```
This provides flexibility but is one additional step. Is there any major drawback for choosing the "master" DC automatically?

trentm replied:
All the operator needs to do here is say "Yes" once. Granted there is an additional cognitive load in having another step. However, my hope is that having to confirm something makes it clear to the operator that there *is* a single DC that is special here: all data needs to be brought over to that one dir for processing.

FWIW, if there is only a single DC, then this step is skipped: the driverDc is automatically set.

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
````diff
@@ -0,0 +1,595 @@
+    ```
+
+- SSH to that server's global zone, and run that script with the postgres
+  VM UUID as an argument. **Run this in screen or equivalent because
````
Would the script fail automatically when a takeover happens during the snaplink discovery? (Reason: The async role could change in the event of a failover so it's important that the process aborts when there is a sitter restart. This will force the operator to rerun the script on that shard.)

trentm replied:
What josh said.

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,595 @@
+After the previous stage, the `mantav2-migrate snaplink-cleanup` command
+will generate a number of "delinking" scripts that must be manually run on
+the appropriate manta service instances. With a larger Manta, expect this
+to be a bit labourious.
```
It'll be great to have the ability to use `manta-oneach` to kick off the script and check back on the progress (operator can always do a small batch at a time). We can consider using `manta-oneach -G -z <vm_uuid1,vm_uuid2,...> '<execute script>'` once the operator has obtained a list of all storage zones which have snaplink objects to be delinked.

trentm replied:
Yes, something like that would be good. I wanted to get something working in here, and then possibly come back to refine steps like that.

One worry I have with automating it somewhat but not completely is that some things could get missed. I don't recall if manta-oneach uses Ur. If so, I have some trust issues there.

Anyway, I think it is a good idea to come back to this to suggest canned oneach commands for the operator of a larger Manta to use.

trentm replied:
MANTA-5142 added to come back to look at doing this

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,595 @@
+    scripts, otherwise metadata will be updated to point to object ids
+    for which no storage file exists.
+
+    There will be one "moraydelink" script for each index moray shard
```
Similar to the comment above, it'll be great to be able to do `manta-oneach -z vm_uuid...` for the list of async instances.

trentm replied:
Agreed. Same answer. MANTA-5142

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,595 @@
+However, there are a couple more steps.
+
+
+### Step 3.5: Update webapi configs and restart
```
Should we add a step to print (for record keeping) and drop the SAPI metadata for ACCOUNTS_SNAPLINKS_DISABLED right before restarting webapi?

trentm replied:
Thanks. Added:
https://github.com/joyent/manta/blob/MANTA-4874/docs/operator-guide/mantav2-migration.md#step-35-remove-the-obsolete-accounts_snaplinks_disabled-metadatum

askfongjojo commented:
Thanks for the write-up. Please see inline comments and some general feedback below:

- All `manta-adm show -a .... | grep ^<service>` can just be `manta-adm show -a .... <service>`
- We probably want to have a step to save all the delink scripts somewhere for record keeping.

joshwilsdon commented on docs/operator-guide/mantav2-migration.md:
````diff
@@ -0,0 +1,595 @@
+    ```
+
+
+## Step 5: Recommended service updates
````
Is the rest of the document expected to be done before this goes to master, or is the plan that these sections will be filled in with a different ticket?

trentm replied:
I'd like to get this in before those sections are complete/polished.  I've updated those sections now (start at https://github.com/joyent/manta/blob/MANTA-4874/docs/operator-guide/mantav2-migration.md#step-5-remove-obsolete-services-and-instances). Any "XXX"s I'd like to hgandle before merging this to master. Any "TODO"s, I'd like to handle separately.

trentm commented:
>     * All `manta-adm show -a .... | grep ^<service>` can just be `manta-adm show -a .... <service>`

Updated, thanks.

>     * We probably want to have a step to save all the delink scripts somewhere for record keeping.

I added a step for that: https://github.com/joyent/manta/blob/MANTA-4874/docs/operator-guide/mantav2-migration.md#step-38-archive-the-snaplink-cleanup-files

trentm commented on docs/operator-guide/mantav2-migration.md:
````diff
@@ -0,0 +1,595 @@
+<!--
+XXX
+```
+mantav2-migrate status
````
I'll drop this. I had been noodling with a `mantav2-migrate status` command that would give state on the other cleanup steps

```
      "MANTA_DELETE_LOG_CLEANUP_REQUIRED": true,
      "REPORTS_CLEANUP_REQUIRED": true,
      "ARCHIVED_JOBS_CLEANUP_REQUIRED": true,
      "MANTAV1_INFRA_CLEANUP_REQUIRED": true
```

but I'll leave this out. I can add details later if anything more is implemented.

trentm commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,595 @@
+When you are sure you have run all these scripts, then answer
+the following to proceed. *WARNING* Be sure you have run all
+these scripts successfully, otherwise lingering snaplinks in the
+system can cause the garbage-collector and rebalancer systems
```
Updated (in commit 034cd2d), thanks.

trentm commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,595 @@
+space those out.
+
+
+### Step 3.6: Tidy up "sherlock" leftovers from stage 3
```
Yes. Updated.

joshwilsdon commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,710 @@
+        The result set size should be exactly 1 (the directory itself).
+
+
+### XXX open Qs
```
Should we remove this open Q's section since we answered the question here?

trentm replied:
Yup, thanks. Dropped.

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
````diff
@@ -0,0 +1,710 @@
+    ...
+    ```
+
+    You can also see the accounts for which Accelerated GC is currently
````
At this point, we are already on the GC v2. The `manta-adm accel-gc` subcommand is gone. The replacement `manta-adm gc` no longer has a concept of whitelisted accounts. Perhaps we can drop this paragraph?

trentm replied:
Ah, thanks!  Dropped.

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,750 @@
+
+There are a number of Manta services that are obsoleted by mantav2 and can
+(and should) be removed at this time. (Note: Remove of these services also
+works before any of the above mantav2 migration steps, if they is easier.)
```
s/they/that/

trentm replied:
done

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,750 @@
+(and should) be removed at this time. (Note: Remove of these services also
+works before any of the above mantav2 migration steps, if they is easier.)
+
+They services (and their instances) to remove are:
```
s/They/The/

trentm replied:
done

trentm commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,750 @@
+## Step 5: Remove obsolete Manta jobs services and instances
+
+There are a number of Manta services that are obsoleted by mantav2 and can
+(and should) be removed at this time. (Note: Remove of these services also
```
done

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
````diff
@@ -0,0 +1,750 @@
+ls -l /var/tmp/snaplink-cleanup*.tgz
+```
+
+And then attach archive that tarball somewhere (perhaps attaching it to your
````
And then attach or archive...

askfongjojo commented on docs/operator-guide/mantav2-migration.md:
```diff
@@ -0,0 +1,750 @@
+They are:
+
+- Clean out the old GC-related `manta_delete_log` (`MANTA_DELETE_LOG_CLEANUP_REQUIRED`).
+- Clean out obsolete reports files under `/:login/reports/' (`REPORTS_CLEANUP_REQUIRED`).
```
`/:login/reports/'  : quote -> backtick
